[LV] Cache call vectorization decisions

huntergr-arm opened the pull request:
LoopVectorize currently queries VFDatabase repeatedly for each CI,
and each query to VFDatabase rescans all vector variants.

This patch instead makes a decision for each call once per VF based
on the cost of scalarization vs. function call to a vector variant
of the function vs. a vector intrinsic, then caches the decision
along with relevant info for use in planning and plan execution.

huntergr-arm committed:
[LV] Cache call vectorization decisions

LoopVectorize currently queries VFDatabase repeatedly for each CI,
and each query to VFDatabase rescans all vector variants.

This patch instead makes a decision for each call once per VF based
on the cost of scalarization vs. function call to a vector variant
of the function vs. a vector intrinsic, then caches the decision
along with relevant info for use in planning and plan execution.

huntergr-arm commented:
Moved this to github to get used to the new workflow.
I believe I've taken care of David's requests for the code, let me know if I missed something.

The changes to using .at() instead of operator[] were because I was performing lookups in a const function, and operator[] is not const (it can create new entries). .at() will assert if the entry does not exist.

I would have split off the other part involving the unit test change, but it appears that the cost was printed out twice and another commit has already updated the test so it no longer fails with my changes.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6970,6 +6968,122 @@ void LoopVectorizationCostModel::setCostBasedWideningDecision(ElementCount VF) {
 }
 
+void LoopVectorizationCostModel::setVectorizedCallDecision(ElementCount VF) {
+  if (VF.isScalar())
```
This function is only called from collectUniformsAndScalars for non-scalar VFs, so I think you can simply assert that the VF is scalar.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1450,6 +1482,7 @@ class LoopVectorizationCostModel {
     if (VF.isScalar() || Uniforms.contains(VF))
       return;
     setCostBasedWideningDecision(VF);
+    setVectorizedCallDecision(VF);
```
This is a comment for line 1479, since github reviews rather helpfully don't let you comment on the actual line you care about. :face_palm

It looks like the comment above collectUniformsAndScalars is out of date. I think it was already out of date before this change, but now it's even worse! Could you add something to the comment about collecting decisions regarding call vectorisation?

huntergr-arm replied:
I've added to the comment above it. To some extent this doesn't feel like the right place to call it, but I found that I needed to call setVectorizedCallDecision() in every place collectUniformsAndScalars() was, so added it. I do think it should only be called on one path though -- currently there's multiple paths where the functions are called, depending on things like a user-supplied VF. I fixed one case, but it seems like it could be simplified with another refactoring patch.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6970,6 +6968,122 @@ void LoopVectorizationCostModel::setCostBasedWideningDecision(ElementCount VF) {
+    for (Instruction &I : *BB) {
+      CallInst *CI = dyn_cast<CallInst>(&I);
+
+      if (!CI || CallWideningDecisions.contains(std::make_pair(CI, VF)))
```
Is it even possible for CallWideningDecisions to already contain CI - would it not be unique for each line of IR?

huntergr-arm replied:
I've removed the check for a preexisting decision.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6970,6 +6968,122 @@ void LoopVectorizationCostModel::setCostBasedWideningDecision(ElementCount VF) {
+          setCallWideningDecision(CI, VF, CM_IntrinsicCall, nullptr,
+                                  getVectorIntrinsicIDForCall(CI, TLI),
+                                  std::nullopt, *RedCost);
+          return;
```
Do you really want to return here? Shouldn't this be `continue` instead? We might have both a fmuladd intrinsic and a sin call in the loop, for example. Is it worth adding a test case for this somehow, since clearly no tests are failing because of it?

huntergr-arm replied:
Yep, good spot. I've added a commit with the fix, I'll add another with a test later.

huntergr-arm committed:
Fix issue with fmuladd reductions cutting widening decisions short

Also documents the extra work done in collectUniformsAndScalars with
this patch, and the code is now stricter about VFs (should never
be called with a scalar VF).

david-arm reviewed (approved):
LGTM! Thanks for making all the changes @huntergr-arm.
